internal/events/aggregator_rewind.go

```diff
@@ -0,0 +1,284 @@
+// Copyright © 2022 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"context"
+	"database/sql/driver"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/firefly/internal/config"
+	"github.com/hyperledger/firefly/internal/data"
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/internal/retry"
+	"github.com/hyperledger/firefly/pkg/database"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+type rewindType int
+
+const (
+	rewindBatch rewindType = iota
+	rewindMessage
+	rewindBlob
+)
+
+type rewind struct {
+	rewindType rewindType
+	uuid       fftypes.UUID
+	hash       fftypes.Bytes32
+}
+
+type rewinder struct {
+	ctx              context.Context
+	aggregator       *aggregator
+	database         database.Plugin
+	data             data.Manager
+	retry            *retry.Retry
+	loop1Done        chan struct{}
+	loop2Done        chan struct{}
+	minRewindTimeout time.Duration
+	mux              sync.Mutex
+	rewindRequests   chan rewind
+	queuedRewinds    []*rewind
+	stagedRewinds    []*rewind
+	loop2ShoulderTap chan bool
+	readyRewinds     map[fftypes.UUID]bool
+}
+
+func newRewinder(ag *aggregator) *rewinder {
+	return &rewinder{
+		ctx:              log.WithLogField(ag.ctx, "role", "aggregator-rewind"),
+		aggregator:       ag,
+		database:         ag.database,
+		data:             ag.data,
+		retry:            ag.retry,
+		loop1Done:        make(chan struct{}),
+		loop2Done:        make(chan struct{}),
+		rewindRequests:   make(chan rewind, config.GetInt(config.EventAggregatorRewindQueueLength)),
+		loop2ShoulderTap: make(chan bool, 1),
+		minRewindTimeout: config.GetDuration(config.EventAggregatorRewindTimeout),
+		readyRewinds:     make(map[fftypes.UUID]bool),
+	}
+}
+
+func (rw *rewinder) start() {
+	go rw.rewindLoop1()
+	go rw.rewindLoop2()
+}
+
+// rewindLoop1 is responsible from taking requests to the aggregator for rewinds, from other
+// event loops that execute completely independently.
+// It takes them in as fast as possible, and puts them in a staging area ready for the aggregator
+// itself to mark it's finished processing anything in-flight
+func (rw *rewinder) rewindLoop1() {
```

Comment:
Would suggest more descriptive names than `rewindLoop1` and `rewindLoop2`... perhaps `receiveLoop` and `processLoop`?

Reply:
Thanks for the suggestions 👍  ... had scratched my head and left these names in out lack of inspiration